// @angular/animations/esm5/src/version.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/**
 * @module
 * @description
 * Entry point for all public APIs of the animation package.
 */
/**
 * @description
 *
 * Represents the version of angular/animations
 */
var /**
 * @module
 * @description
 * Entry point for all public APIs of the animation package.
 */
/**
 * @description
 *
 * Represents the version of angular/animations
 */
Version = /** @class */ (function () {
    function Version(full) {
        this.full = full;
        var splits = full.split('.');
        this.major = splits[0];
        this.minor = splits[1];
        this.patch = splits.slice(2).join('.');
    }
    return Version;
}());
/**
 * @module
 * @description
 * Entry point for all public APIs of the animation package.
 */
/**
 * @description
 *
 * Represents the version of angular/animations
 */
export { Version };
/**
 *
 */
export var VERSION = new Version('6.0.0');
